src/common/media.cpp

```diff
@@ -0,0 +1,172 @@
+            // longer than 500ms
+            if ((tick - timer) > 500) {
+                _dbg("USB host reenumerating"); // trace
+                USBH_ReEnumerate(&hUsbHostHS);  // re-enumerate UsbHost
+            }
+        } else // otherwise update timer
+            timer = tick;
+    }
+}
+
+uint8_t media_inserted = 0;
+media_error_t media_error = media_error_OK;
+media_print_state_t media_print_state = media_print_state_NONE;
+char media_print_filename[128] = { 0 };
+char media_print_filepath[128] = { 0 };
+FIL media_print_fil;
+uint32_t media_current_position;
+uint32_t media_current_line;
+uint32_t media_queue_position[BUFSIZE];
+uint32_t media_queue_line[BUFSIZE];
+
+uint8_t media_is_inserted(void) {
+    return media_inserted;
+}
+
+void media_print_start(const char *filepath) {
+    FILINFO filinfo;
+    if (media_print_state == media_print_state_NONE) {
+        strncpy(media_print_filepath, filepath, sizeof(media_print_filepath) - 1);
+        if (f_stat(media_print_filepath, &filinfo) == FR_OK) {
+            strncpy(media_print_filepath, filinfo.fname, sizeof(media_print_filepath) - 1);
+            if (f_open(&media_print_fil, media_print_filepath, FA_READ) == FR_OK) {
+                media_current_position = 0;
+                media_current_line = 0;
+                media_print_state = media_print_state_PRINTING;
+            }
+        }
+    }
+}
+
+void media_print_stop(void) {
+    if ((media_print_state == media_print_state_PRINTING) || (media_print_state == media_print_state_PAUSED)) {
+        f_close(&media_print_fil);
+        queue.clear();
+        media_print_state = media_print_state_NONE;
+    }
+}
+
+void media_print_pause(void) {
+    if (media_print_state == media_print_state_PRINTING) {
+        f_close(&media_print_fil);
+#if 0
+		int length = queue.length;
+		int index_r = queue.index_r + length - 1;
+		if (index_r >= BUFSIZE)
+			index_r -= BUFSIZE;
+		while (length--)
+		{
+			media_current_position = media_queue_position[index_r];
+			media_current_line = media_queue_line[index_r];
+			_dbg("dequeue %5u %s", media_current_line, queue.command_buffer[index_r]);
+			if (--index_r < 0)
+				index_r = BUFSIZE - 1;
+		}
+#else
+        int index_r = queue.index_r;
+        media_current_position = media_queue_position[index_r];
+        media_current_line = media_queue_line[index_r];
+        _dbg("dequeue %5u %s", media_current_line, queue.command_buffer[index_r]);
+#endif
+
+        queue.clear();
+        media_print_state = media_print_state_PAUSED;
+    }
+}
+
+void media_print_resume(void) {
+    if (media_print_state == media_print_state_PAUSED) {
+        if (f_open(&media_print_fil, media_print_filepath, FA_READ) == FR_OK) {
+            if (f_lseek(&media_print_fil, media_current_position) == FR_OK)
+                media_print_state = media_print_state_PRINTING;
+            else
+                f_close(&media_print_fil);
+        }
+    }
```

Comment:
Again, what if any of these functions fail?

Reply:
I think this is clear - media_print_state stay unchanged and user can repeat resume action.